SubsampledNeighborsTransformer: Subsampled nearest neighbors for faster and more space efficient estimators that accept precomputed distance matrices

jenniferjang opened the pull request:
<!--
Thanks for contributing a pull request! Please ensure you have taken a look at
the contribution guidelines: https://github.com/scikit-learn/scikit-learn/blob/master/CONTRIBUTING.md#pull-request-checklist
-->

#### Reference Issues/PRs
<!--
Example: Fixes #1234. See also #3456.
Please use keywords (e.g., Fixes) to create link to the issues or pull requests
you resolved, so that they will automatically be closed when your pull request
is merged. See https://github.com/blog/1506-closing-issues-via-pull-requests
-->
See #17650


#### What does this implement/fix? Explain your changes.
Instead of calculating the pairwise distances for all pairs of points to obtain nearest-neighbor graphs for estimators like DBSCAN, SubsampledNeighborsTransformer only calculates distances for a fraction s of the pairs (selected uniformly at random). This would make estimators that accept precomputed distance matrices feasible for larger datasets. In a very recent work I did with Google Research [1], we found that you can get over 200x speedup and 250x savings in memory this way without hurting the clustering quality (in some cases s = 0.001 suffices).

[1] https://arxiv.org/abs/2006.06743

#### Any other comments?


<!--
Please be aware that we are a loose team of volunteers so patience is
necessary; assistance handling other issues is very welcome. We value
all user contributions, no matter how minor they are. If we are slow to
review, either the pull request needs some benchmarking, tinkering,
convincing, etc. or more likely the reviewers are simply busy. In either
case, we ask for your understanding during the review process.
For more information, see our FAQ on this topic:
http://scikit-learn.org/dev/faq.html#why-is-my-pull-request-not-getting-any-attention.

Thanks for contributing!
-->

thomasjpfan commented on sklearn/neighbors/_subsampled.py:
```diff
@@ -0,0 +1,193 @@
+    and the final distance matrix is symmetric.
+    """
+
+    @_deprecate_positional_args
```
Since this is a new estimator, we do not need to do a deprecation.

jenniferjang replied:
Ah, okay.

thomasjpfan commented on sklearn/neighbors/_subsampled.py:
```diff
@@ -0,0 +1,193 @@
+    Parameters
+    ----------
+
+    s : float
```
Is there a default value we can use here? If not how would a user decide on what `s` they should use?

jenniferjang replied:
I've added the default s=0.1 here, since that was a setting we found worked for many datasets.

thomasjpfan commented on sklearn/neighbors/_subsampled.py:
```diff
@@ -0,0 +1,193 @@
+        if self.metric not in PAIRED_DISTANCES and not callable(self.metric):
+            raise ValueError('Unknown distance %s' % self.metric)
+
+        self.eps_ = self.eps
+        self.s_ = self.s
+        self.metric_ = self.metric
```
We can use the parameters directly and not define new attributes.

```suggestion
```

thomasjpfan commented on sklearn/neighbors/_subsampled.py:
```diff
@@ -0,0 +1,193 @@
+        self.random_state = random_state
+
+
+    def _fit(self, X):
```
```suggestion
    def _check_parameters(self, X):
```

jenniferjang replied:
I moved the checking to a **_check_parameters** function.

thomasjpfan commented on sklearn/neighbors/__init__.py:
```diff
@@ -15,6 +15,7 @@
 from ._kde import KernelDensity
 from ._lof import LocalOutlierFactor
 from ._nca import NeighborhoodComponentsAnalysis
+from ._subsampled import SubsampledNeighborsTransformer
```
This should be added to `__all__`.

jenniferjang replied:
added!

thomasjpfan reviewed (commented):
Thank you for the PR @jenniferjang !

You can run `make flake8-diff` locally to find the flake8 errors.

Are there references of this approach being used before https://arxiv.org/abs/2006.06743 ?

jenniferjang commented:
> Thank you for the PR @jenniferjang !
> 
> You can run `make flake8-diff` locally to find the flake8 errors.
> 
> Are there references of this approach being used before https://arxiv.org/abs/2006.06743 ?

Hi @thomasjpfan, thanks for reviewing this. I initially implemented it as a transformer, but do you think it would work better as  a function instead? I can't get the test check_methods_subset_invariance to pass for the transformer because in our case, transform generates the distance matrix for the data, which shouldn't satisfy the subset invariance property.

To my knowledge I haven't seen this approach being used before in the literature, at least in the context of DBSCAN.

jnothman commented:
>  I can't get the test check_methods_subset_invariance to pass for the transformer because in our case, transform generates the distance matrix for the data, which shouldn't satisfy the subset invariance property.

It's not because of the distance matrix that it shouldn't satisfy the subset invariance property, only because of the random nature of the transformation. See other estimators that disable 'check_methods_subset_invariance', such as `DummyClassifier` and `BernoulliRBM`

jenniferjang commented:
> > I can't get the test check_methods_subset_invariance to pass for the transformer because in our case, transform generates the distance matrix for the data, which shouldn't satisfy the subset invariance property.
> 
> It's not because of the distance matrix that it shouldn't satisfy the subset invariance property, only because of the random nature of the transformation. See other estimators that disable 'check_methods_subset_invariance', such as `DummyClassifier` and `BernoulliRBM`

I've added the tag to skip the test.

jenniferjang commented:
@thomasjpfan @jnothman Thanks for the speedy review! I've updated the fit function to better conform to the transformer paradigm. Please take a look at the updated code at your convenience.

jenniferjang commented:
@thomasjpfan @jnothman Good morning, can you please take a look at the changes? Thank you!

thomasjpfan commented:
Sorry for the late response. This PR most likely needs an example to demonstrate how it compares to regular DBSCAN. Looking at your paper a good candidate for an example are the Faces or Bank dataset. We do not want the example to run for too long because we run them every time the docs get built. The example can also cover how `s` is determined.

cmarmo commented:
Hi @jenniferjang , thanks for your work. Please note that some checks are failing because of `sklearn.neighbors._base` missing the attribute 'UnsupervisedMixin'.... I can't understand why I can't find the class in the github diff, but I'm finding it when checking out your branch... Do you mind checking if something went wrong in the workflow? Thanks!

jenniferjang commented:
Sorry for the late response! @jnothman, @thomasjpfan, @cmarmo, it appears that the class `UnsupervisedMixin`, which was one of the parent classes of SubsampledNeighborsTransformer, was removed from scikit-learn, which caused the errors. I've removed dependencies on `UnsupervisedMixin`, and I believe it builds now. 

I’ve also added a new example, `plot_subsampled_neighbors_transformer_dbscan.py`, which plots DBSCAN and subsampled DBSCAN results side by side. However this seem to have issues building, and I'm not sure how to fix it.

One thing I noticed is that paired_distances is awfully inefficient: on 30,000 points, `paired_distances` alone on 10% of edges took more than twice as much time as it took to run the entirety of DBSCAN without sampling. Depending on the size of the dataset, `paired_distances` takes between 50%-80% of the total runtime of subsampled DBSCAN. I’d like to look into improving `paired_distances`, either before or after finishing this pull request. Otherwise it would be infeasible to use `SubsampledNeighborsTransformer` to speed up DBSCAN unless the dataset is extremely large.

In order to speed up subsampled DBSCAN, I sort the output of `SubsampledNeighborsTransformer`’s fit_transform. If you see other ways to make the `subsampled_neighbors` function faster, please let me know.

thomasjpfan commented:
> One thing I noticed is that paired_distances is awfully inefficient: on 30,000 points, paired_distances alone on 10% of edges took more than twice as much time as it took to run the entirety of DBSCAN without sampling.

May you see if `pairwise_distances_chunked` would improve the performance?

jenniferjang commented:
> > One thing I noticed is that paired_distances is awfully inefficient: on 30,000 points, paired_distances alone on 10% of edges took more than twice as much time as it took to run the entirety of DBSCAN without sampling.
> 
> May you see if `pairwise_distances_chunked` would improve the performance?

I looked into pairwise_distances_chunked. It calculates the distance matrix for all pairs of points, right? In that case our performance would be n^2 -- the same as DBSCAN -- and wouldn't work for large inputs. Is there a way for `pairwise_distances_chunked` to calculate distances for a different (random) subset of neighbors per point? I looked into using `reduce_func` but as far as I saw, that would be applied after the full distance matrix is calculated.

jnothman commented:
Interesting. The pairwise_distances implementation of Euclidean distance is
taking advantage of having a constant norm to use in calculating each row
and column of the distance matrix.

I don't think you can use chunking here in any straightforward way. You
would need to implement a version of paired distances that accepted norms
for each X,Y, as with the fast Euclidean pairwise distances implementation.

iqbalfarz commented:
Hi @jenniferjang, any update on this?

jenniferjang commented:
New PR: https://github.com/scikit-learn/scikit-learn/pull/30523
